thirdeye/thirdeye-pinot/src/main/java/org/apache/pinot/thirdeye/datasource/cache/metadata/MetaDataCollectionCache.java

```diff
@@ -0,0 +1,83 @@
+/*
+ *  Licensed to the Apache Software Foundation (ASF) under one
+ *  or more contributor license agreements.  See the NOTICE file
+ *  distributed with this work for additional information
+ *  regarding copyright ownership.  The ASF licenses this file
+ *  to you under the Apache License, Version 2.0 (the
+ *  "License"); you may not use this file except in compliance
+ *  with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing,
+ *  software distributed under the License is distributed on an
+ *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  KIND, either express or implied.  See the License for the
+ *  specific language governing permissions and limitations
+ *  under the License.
+ *
+ */
+
+package org.apache.pinot.thirdeye.datasource.cache.metadata;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The cache for a meta-data collection.
+ * This cache will hold a collection of meta-data and periodically reloads it using the cache loader.
+ *
+ * It's designed to cache the original or derived meta-data collection for ThirdEye Dashboard server, for example, a
+ * list of all applications, subscription groups, and detection configs. Loading a collection of metadata has been
+ * a pain point for ThirdEye. It requires data base to scan the index table and then do a look up for each config in the
+ * generic json table, which is slow. This cache makes such loading asynchronous and returns user the cached result
+ * when opening page.
+ *
+ * @param <T> the type of the value to be cached.
+ */
+public class MetaDataCollectionCache<T> {
+  private static final String META_DATA_KEY = "metaData";
+  protected static final Logger LOG = LoggerFactory.getLogger(MetaDataCollectionCache.class);
+  private final LoadingCache<String, T> loadingCache;
+
+  /**
+   * Create a Metadata collection cache
+   * @param cacheLoader the cache loader for the meta data collection
+   * @param reloadPeriod the period of the reloading frequency
+   * @param reloadUnit the unit for the reload period
+   */
+  public MetaDataCollectionCache(MetadataCacheLoader<T> cacheLoader, int reloadPeriod, TimeUnit reloadUnit) {
+    this.loadingCache = CacheBuilder.newBuilder()
+        .expireAfterWrite(10, TimeUnit.MINUTES)
```

Comment:
This looks dangerous.
What if user disabled/enabled the config, and then refresh the page?
Will the user see the updated config?

Reply:
Good point. Yes, it should. It will trigger a cache refresh whenever the user changes the config.

Reply:
Config could be changed in many places.
Are you sure the cache is always updated?

Reply:
It should be. It will trigger a cache refresh whenever 1. Detection or Subscription group YAML change. 2. Toggle the active button.  Also, there is a thread that refreshes the cache every 3 minutes.  The cache is kept very fresh. 

The goal of the cache is to pre-load the list into memory so that users don't need to wait for the results from the database when opening the page. In my testing, this reduces the page latency from around 3s to around 0.3s.

Reply:
Although this looks great, we need to keep in mind that whenever we update the metadata in the future, we need to explicitly make a call to cache refresh which is not very trivial and if we miss this, then it can be a pain while debugging.

Can we make the cache refresh call part of the data access layer?

Reply:
Actually, this cache is only used in the UI endpoint to show the list of alerts. Most of the time, it's not necessary to update the cache whenever we update the metadata because that page doesn't show that information and the consistency requirement is not very high. The most important places to refresh it are creating alerts, activate/deactivate the alerts, create/update subscription groups, where users expect to see the changes in UI. Otherwise, I think an eventual consistency model would be enough, it will be refreshed in 3 minutes